File: 1398-number-of-ways-to-stay-in-the-same-place-after-some-steps/1398-number-of-ways-to-stay-in-the-same-place-after-some-steps.java
public class Solution {
    private static final int MOD = 1000000007;

    public int numWays(int steps, int arrLen) {
        int[][] dp = new int[steps + 1][Math.min(steps, arrLen) + 2];
        dp[0][0] = 1;

        for (int step = 1; step <= steps; step++) {
            for (int pos = 0; pos < Math.min(steps, arrLen); pos++) {
                dp[step][pos] = (((dp[step - 1][pos] % MOD)  // stay
                        + (dp[step - 1][pos + 1] % MOD)) % MOD  // right
                        + ((pos > 0) ? (dp[step - 1][pos - 1] % MOD) : 0)) % MOD;  // left
            }
        }
        return dp[steps][0];
    }
}

File: 1398-number-of-ways-to-stay-in-the-same-place-after-some-steps/1398-number-of-ways-to-stay-in-the-same-place-after-some-steps.js
var numWays = function (steps, arrLen) {
    const MOD = 10 ** 9 + 7;
    let dp = Array.from({ length: steps + 1 }, () => Array(Math.min(steps, arrLen) + 2).fill(0));
    dp[0][0] = 1;

    for (let step = 1; step <= steps; step++) {
        for (let pos = 0; pos < Math.min(steps, arrLen); pos++) {
            dp[step][pos] = ((dp[step - 1][pos] + dp[step - 1][pos + 1]) % MOD + (pos > 0 ? dp[step - 1][pos - 1] : 0)) % MOD;
        }
    }
    return dp[steps][0];
};